flang/lib/Evaluate/tools.cpp

```diff
@@ -1203,6 +1203,10 @@ bool HasVectorSubscript(const Expr<SomeType> &expr) {
   return HasVectorSubscriptHelper{}(expr);
 }
 
+bool IsArraySection(const Expr<SomeType> &expr) {
+  return expr.Rank() > 0 && IsVariable(expr) && !UnwrapWholeSymbolDataRef(expr);
```

Comment:
I think that this predicate would be true if the expr were a whole symbol data ref to a non-`CONTIGUOUS` assumed-shape dummy array or non-`CONTIGUOUS` pointer, and while those are not "sections" strictly speaking, they are going to get copied in/out.